search/agent/base.py

```diff
@@ -0,0 +1,182 @@
+
+    This class is instantiated when the containing script is run by the
+    MultiLangDaemon. The underlying KCL process handles threading, offset
+    tracking, etc. The KCL guarantees that our :class:`.RecordProcessor` will
+    see each record *at least* once; the checkpoint mechanism gives us a way to
+    further guarantee that we only process each record a single time.
+
+    See the ``consumer.properties`` file for configuration details, including
+    streams.
+    """
+
+    def __init__(self) -> None:
+        """Initialize checkpointing state and retry configuration."""
+        self._SLEEP_SECONDS = 5
+        self._CHECKPOINT_RETRIES = 5
+        self._CHECKPOINT_FREQ = 60
+        self._largest_seq: Tuple[Optional[int], Optional[int]] = (None, None)
+        self._largest_sub_seq = None
+        self._last_checkpoint_time: Optional[float] = None
+
+    def initialize(self, initialize_input: Any) -> None:
+        """Called once by a KCLProcess before any calls to process_records."""
+        self._largest_seq = (None, None)
+        self._last_checkpoint_time = time.time()
+
+    def checkpoint(self, checkpointer: amazon_kclpy.kcl.Checkpointer,
+                   sequence_number: Optional[str]=None,
+                   sub_sequence_number: Optional[int]=None) -> None:
+        """Make periodic checkpoints while processing records."""
+        for n in range(0, self._CHECKPOINT_RETRIES):
+            try:
+                checkpointer.checkpoint(sequence_number, sub_sequence_number)
+                return
+            except kcl.CheckpointError as e:
+                if e.value == 'ShutdownException':
+                    # A ShutdownException indicates that this record processor
+                    #  should be shutdown. This is due to some failover event,
+                    #  e.g. another MultiLangDaemon has taken the lease for
+                    #  this shard.
+                    logger.info("Encountered shutdown exception, skipping"
+                                " checkpoint")
+                    return
+                elif e.value == 'ThrottlingException':
+                    # A ThrottlingException indicates that one of our
+                    #  dependencies is is over burdened, e.g. too many dynamo
+                    #  writes. We will sleep temporarily to let it recover.
+                    if self._CHECKPOINT_RETRIES - 1 == n:
+                        logger.error("Failed to checkpoint after %i attempts,"
+                                     " giving up.", n)
+                        return
+                    else:
+                        logger.info("Was throttled while checkpointing, will"
+                                    " attempt again in %i seconds",
+                                    self._SLEEP_SECONDS)
+                elif e.value == 'InvalidStateException':
+                    logger.error("MultiLangDaemon reported an invalid state"
```

Comment:
Nose tests seem to be getting this error https://travis-ci.org/cul-it/arxiv-search/builds/351393938

Reply:
Some of the tests are for error conditions, so if the test is passing that's likely what's going on.